### ISSUE TITLE
"per-resource" run mode queries kubernetes metrics with wrong labelSelector

### ISSUE THREAD

**Issue by StefanCenusa** [jthomperoo/custom-pod-autoscaler#158]
**Describe the bug**

I've run into an issue when trying to write a custom-pod-autoscaler that uses kubernetes metrics.
The CPA logged `Error while autoscaling: failed to get metrics: failed to get required Kubernetes metrics: invalid metrics (1 invalid out of 1), first error is: failed to get resource metric: unable to get metrics for resource cpu: no metrics returned from resource metrics API` although metrics were available for that resource. It only worked after I've added `labels` on my deployment resource that matched the pod labels.

I think this issue is caused by [this code](https://github.com/jthomperoo/custom-pod-autoscaler/blob/8f2cd13c7db2969964d65c8fd57eb34f6bfbfcaa/internal/k8smetricget/k8smetricget.go#L115) where CPA tries to get metrics based on the resource's labels instead of the ones defined in `spec.selector` of a Deployment or a StatefulSet.
```
metric, err := c.getMetric(currentReplicas, spec, namespace, labels.Set(resource.GetLabels()).AsSelector())
```

In my opinion, this behaviour is unwanted, because there are cases where Deployments could have different labels than pods they own. Thus, querying for Pods' metrics with labelSelector built from Deployment's labels could lead to no metrics being returned. 

**To Reproduce**
1. Deploy a Deployment with no labels defined
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: hello-kubernetes
spec:
  replicas: 1
  selector:
    matchLabels:
      app: hello-kubernetes
  template:
    metadata:
      labels:
        app: hello-kubernetes
    spec:
      containers:
 ...
```
2. Create a CPA that uses kubernetes metrics; use `runMode: "per-resource"`
3. Target the CPA to this deployment
```
...
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: hello-kubernetes
  roleRequiresMetricsServer: true
...
```
4. Enabled verbose logs on CPA
5. See error

**Expected behavior**
CPA to query for metrics based on Pods of the resource, not the resource itself when it is a Deployment or StatefulSet.

**Comment by jthomperoo** [comment]
Thanks very much for finding this and providing a really clear set of steps to recreate and detail the issue!

Sorry it took a while to get this fixed, just haven't had time recently.

As a bonus effect of fixing this now the CPA supports [match
expressions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#resources-that-support-set-based-requirements) for selecting K8s metrics, which the previous incorrect implementation did not support.

This is now available in release [`v2.4.1`](https://github.com/jthomperoo/custom-pod-autoscaler/releases/tag/v2.4.1).

Thanks!

**Comment by StefanCenusa** [comment]
Hello, @jthomperoo! Thanks for fixing this.
